packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts

```diff
@@ -0,0 +1,1102 @@
+import _ from 'lodash';
+import {
+  convertFuncAppToElem,
+  createDummyOperand,
+  isAsKeyword,
+  isDummyOperand,
+  markInvalid,
+} from './utils';
+import { CompileError, CompileErrorCode } from '../errors';
+import { SyntaxToken, SyntaxTokenKind, isOpToken } from '../lexer/tokens';
+import Report from '../report';
+import { ParsingContext, ParsingContextStack } from './contextStack';
+import {
+  AttributeNode,
+  BlockExpressionNode,
+  CallExpressionNode,
+  ElementDeclarationNode,
+  FunctionApplicationNode,
+  FunctionExpressionNode,
+  GroupExpressionNode,
+  IdentiferStreamNode,
+  InfixExpressionNode,
+  ListExpressionNode,
+  LiteralNode,
+  NormalExpressionNode,
+  PostfixExpressionNode,
+  PrefixExpressionNode,
+  PrimaryExpressionNode,
+  ProgramNode,
+  SyntaxNode,
+  SyntaxNodeIdGenerator,
+  TupleExpressionNode,
+  VariableNode,
+} from './nodes';
+import NodeFactory from './factory';
+import { hasTrailingNewLines, hasTrailingSpaces, isAtStartOfLine } from '../lexer/utils';
+
+/* eslint-disable no-loop-func */
+
+// A class of errors that represent a parsing failure and contain the node that was partially parsed
+class PartialParsingError<T extends SyntaxNode | undefined> {
```

Comment:
How about making `partialNode` optional instead of extends `undefined`? Personally, I find it easier to read